fedmsg_migration_tools/verify_missing.py

```diff
@@ -0,0 +1,168 @@
+        self._socket = None
+        self._factory = None
+
+    def startService(self):
+        self._factory = ZmqFactory()
+        endpoints = [
+            ZmqEndpoint(ZmqEndpointType.connect, endpoint)
+            for endpoint in self.endpoints
+        ]
+        _log.debug('Configuring ZeroMQ subscription socket')
+        for endpoint in endpoints:
+            s = ZmqSubConnection(self._factory, endpoint)
+            s.subscribe(b"")
+            s.gotMessage = self.on_message
+        _log.info('ZeroMQ consumer is ready')
+
+    def on_message(self, body, topic):
+        topic = topic.decode("utf-8")
+        msg = json.loads(body)
+        _log.debug('Received from ZeroMQ on topic %s: %s',
+                   topic, msg["msg_id"])
+        self.store[msg["msg_id"]] = (
+            datetime.utcnow(),
+            msg,
+        )
+
+    def stopService(self):
+        _log.debug("Stopping ZmqConsumer")
+        self._factory.shutdown()
+
+
+class Comparator(service.Service):
+
+    MATCH_WINDOW = 20
+
+    def __init__(self, amqp_store, zmq_store):
+        self.amqp_store = amqp_store
+        self.zmq_store = zmq_store
+        self._rm_loop = task.LoopingCall(self.remove_matching)
+        self._cm_loop = task.LoopingCall(self.check_missing)
+
+    def startService(self):
+        self._rm_loop.start(1)
+        self._cm_loop.start(10)
+
+    def stopService(self):
+        _log.debug("Stopping Comparator")
+        self._rm_loop.stop()
+        self._cm_loop.stop()
+
+    def remove_matching(self):
+        _log.debug("Checking for matching messages (%i, %i)",
+                   len(self.amqp_store), len(self.zmq_store))
+        for msg_id in list(self.amqp_store.keys()):
+            if msg_id in self.zmq_store:
+                del self.amqp_store[msg_id]
+                del self.zmq_store[msg_id]
+
+    def check_missing(self):
+        _log.debug("Checking for missing messages")
+        self._check_store(self.amqp_store, "AMQP")
+        self._check_store(self.zmq_store, "ZeroMQ")
+
+    def _check_store(self, store, source_name):
+        threshold = datetime.utcnow() - timedelta(seconds=self.MATCH_WINDOW)
+        for msg_id, value in list(store.items()):
+            time, msg = value
+            if time < threshold:
+                _log.warning("Message %s was only received in %s",
+                             msg_id, source_name)
+                del store[msg_id]
+
+
+def main():
+    amqp_store = {}
+    zmq_store = {}
+    verify_service = MasterService()
+    comparator = Comparator(amqp_store, zmq_store)
+    comparator.setServiceParent(verify_service)
+    zmq_consumer = ZmqConsumer(zmq_store)
+    zmq_consumer.setServiceParent(verify_service)
+    amqp_consumer = AmqpConsumer(amqp_store)
+    amqp_consumer.setServiceParent(verify_service)
+    verify_service.startService()
+    try:
+        reactor.run()
+    except KeyboardInterrupt:
```

Comment:
I believe there's a signal handler installed by fedora-messaging to handle SIGINT and SIGTERM. We should make sure it's easy to control when that's installed so you could override it here with your own handler to perform additional shutdown tasks.

Reply:
Right, but it's only in the `ConsumerSession` class and I'm not using that: it' synchronous and only intended for the `consume` api function.

Reply:
Ah right. I've got a memory like a sieve.